File: 31.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        if len(nums)<=1:
            nums[:] = nums
        else:
            left = len(nums) - 1
            while left > 0 and nums[left] <=  nums[left - 1]:
                left -= 1
            if left == 0:
                return self.reverse(nums, 0, len(nums) - 1)
            left -= 1
            right = 0
            for i in range(len(nums)-1, left, -1):
                if nums[i] > nums[left]:
                    right = i
                    break
            nums[left], nums[right] = nums[right], nums[left]
            self.reverse(nums, left + 1, len(nums)-1)
    def reverse(self, nums, l, r):
        while l<r:
            nums[l], nums[r] = nums[r], nums[l]
            l += 1
            r -= 1

File: 31.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int index = nums.size() - 1, next = 0;
        for(int i = nums.size() - 2; i>= 0; i--){
            if (nums[i] < nums[i+1]){
                index = i;
                break;
            }
        }
        if (index == nums.size() - 1){
            std::sort(nums.begin(), nums.end());
        }else{
            next = index + 1;
            while((next + 1) < nums.size() and nums[index] < nums[next+1]){
                next++;
            }
            int tmp = nums[next];
            nums[next] = nums[index];
            nums[index] = tmp;
            std::sort(nums.begin() + index + 1, nums.end());
        }

    }
};
